mobilecoind/src/sync.rs

```diff
@@ -286,8 +296,118 @@ fn sync_thread_entry_point(
     }
 }
 
-/// Sync a single monitor.
-fn sync_monitor(
+/// Sync a single monitor, spawning parallel jobs to perform block scanning if
+/// it is many blocks behind. This allows us to use all the CPU cores scanning
+/// even when we only have a single monitor to work on, and all the blocks are
+/// pretty small (only 2 or 3 UTXOs), which is a very common configuration and
+/// situation. First, we test how far behind the monitor is. When the monitor is
+/// close to caught up, we fallback to sync_monitor_sequential.
+fn sync_monitor_parallel(
+    ledger_db: &LedgerDB,
+    mobilecoind_db: &Database,
+    monitor_id: &MonitorId,
+    num_workers: usize,
+    logger: &Logger,
+) -> Result<SyncMonitorOk, Error> {
+    let monitor_data = mobilecoind_db.get_monitor_data(monitor_id)?;
+    let num_blocks = ledger_db.num_blocks()?;
+
+    // If we cannot hope to work on num_workers blocks in parallel for this montior,
+    // then we are pretty close to the end, let's fall back to
+    // sync_monitor_sequential. Otherwise, we should be able to get num_workers
+    // blocks and work on them in parallel.
+    if monitor_data.next_block + num_workers as u64 >= num_blocks {
+        return sync_monitor_sequential(ledger_db, mobilecoind_db, monitor_id, logger);
```

Comment:
It might also be reasonable to skip this fallback, and make this the only version of sync_monitor. I don't think it's going to change things that much, since if you are at the end, then n-1 of the workers will immediately get a block not found error and then stop working on this and do something else, since rayon works via thread pools.

OTOH that is a less conservative change than this one. For this one, once the monitors are caught up, we aren't changing behavior at all, the new code is basically only running during the initial catch-up period.

Reply:
If we did that though, we would have to make the parallel version sensitive to the lefger db not found error, which is a more complex change. On balance i think having two versions is simpler because of that.

Reply:
Can't you cap num_workers at `num_blocks - monitor_data.next_block`?

Reply:
Hmm -- I am thinking through what will happen to the corresponding logic from the sequential version here:

```
        let block_contents = match ledger_db.get_block_contents(monitor_data.next_block) {
            Ok(block_contents) => block_contents,
            Err(mc_ledger_db::Error::NotFound) => {
                return Ok(SyncMonitorOk::NoMoreBlocks);
            }
            Err(err) => {
                return Err(err.into());
            }
        };
```

Currently the parallel version never returns `SyncMonitorOk::NoMoreBlocks`, because it always calls to the sequential version if we are close to the end.

If we did the change you are saying, then I guess when num_blocks == monitor_data.next_block (there is no work to do), then num_workers will be zero and all these loops will pass right through.

So then this line:

```
    Ok(SyncMonitorOk::MoreBlocksPotentiallyAvailable)
```

Should be replaced with something like:

```
  if num_cpus > num_blocks - monitor_data.next_block {
    Ok(SyncMonitorOk::MoreBlocksPotentiallyAvailable)
  } else {
    Ok(SyncMonitorOk:NoMoreBlocks)
  }
```

or something like this? Does that sound right?

I'm not sure I fully worked through the details but if this sounds better to you, then we can explore it.

Reply:
Or maybe if monitor_data.next_block > num_blocks we early-return SyncMonitorOk:NoMoreBlocks?

Reply:
Okay, I will try changing it the way you suggest